string/string.h

```diff
@@ -0,0 +1,297 @@
+
+        for (int i = strlen + 2; i > position - 1; i--) {
+            this->string[i] = this->string[i - 1];
+        }
+
+        this->string[position] = character;
+        this->strlen++;
+    }
+
+    /**
+     * Removes character at position specified
+     * @param position to remove at
+     */
+    void remove(int position) {
+        if (position > strlen) return;
+
+        for (int i = position; i < strlen + 1; i++) {
+            this->string[i] = this->string[i + 1];
+        }
+        this->strlen--;
+    }
+
+    /**
+     * Wipes the entire string buffer
+     */
+    void clear() {
+        int i = 0;
+        while (this->strlen > 0) {
+            string[i++] = NULL_TERMINATOR;
+            this->strlen--;
+        }
+    }
+
+    /**
+     * Recount the length of the string in case
+     * null gets randomly added to a string
+     *
+     * @return New Length or -1 if unchanged
+     */
+    int recount() {
+        int len_count = 0;
+
+        while (this->string[(len_count++) + 1] != NULL_TERMINATOR);
+
+        if (len_count != this->strlen) {
+            this->strlen = len_count;
+
+            return this->strlen;
+        }
+
+        return -1;
+    }
+
+    /**************************************
+     * Iterators
+     **************************************/
+    // TODO: Could be nice. Maybe add later after getting base functionality down.
+
+    /**************************************
+     * Capacity
+     **************************************/
+
+    /**
+     * Check if string is empty
+     * @return
+     */
+    bool empty() const {
+        return this->strlen == 0;
+    }
+
+    /**
+     * Check if string fills entire buffer
+     * @return
+     */
+    bool full() const {
+        return this->strlen >= size;
+    }
+
+    /**
+     * Get the max size of the buffer
+     * @return
+     */
+    size_t max_size() const {
+        return size;
+    }
+
+    /**
+     * Get the remaining available space in the buffer
+     * @return
+     */
+    size_t available() const {
+        return size - strlen - 1;
+    }
+
+    /**************************************
+     * C API Interfacing
+     **************************************/
+
+    /**
+     * Checks to see if the string buffer is full
+     *
+     * @return boolean on if there buffer is filled
+     */
+    bool is_full() {
+        return this->strlen >= size;
+    }
+
+    /**
+     * Checks if there is extra space in the buffer
+     * @return boolean on if there's extra space
+     */
+    bool is_trunc() {
+        return this->strlen < size;
+    }
+
+    /**
+     * Get the first terminator
+     * @return
+     */
+    char *get_end() {
+        return this->string[strlen];
+    }
+
+protected:
+    String(char const *buffer) {
+        int i; // Reduce errors from using strlen as counter
+
+        for (i = 0; i < size - 1; i++) {
+            this->string[i] = buffer[i];
```

Comment:
I think this is where the confusion is, you shouldn't be copying 'buffer' to your own buffer, that still means you need to allocate your own memory and we don't want to specify how much we need to allocate in this class. In this constructor you should be making a 'string' a pointer and setting it to the address held by 'buffer', you should also pass in the size of buffer so you know how big it is. This way you have derived classes which are templated by size and their only job is to allocate memory for this parent class, which then just uses that memory without every needing to allocate it's own.

Reply:
The end goal is that we need to be able to pass strings around without specifying the size.
This function would only work on strings of length 10 even though we likely don't actually care how big it is
`void foo(String<10> name);` 

In the general case you would template the function, which means you're generating a function for every single length of string you call it on AND you have to do that all at compile time!

> template typename\<const size_t SIZE\>
> void foo(String\<SIZE\> name);

Instead we would rather use a function like this `void foo(String name);` which will work on strings of whatever size you want. But we still need to be able to allocate space for it, but this function doesn't care, only whoever declared the string in the first place cares. That's why we create our strings as `alloc::String<10> string` and then pass them around as generic `String` types.